File: src/api/src/services/llm/function_calling/stealth.py
"""Stealth HTTP utilities for anti-bot evasion.

Provides browser-like request profiles, user-agent rotation, proxy support,
random delays, and FlareSolverr integration.

Configuration (environment variables):
  RECALL_PROXY_URL           - Single residential proxy URL
  RECALL_PROXY_URLS          - Comma-separated proxy URLs for rotation
  RECALL_FLARESOLVERR_URL    - FlareSolverr endpoint (e.g. http://localhost:8191)
  RECALL_STEALTH_MIN_DELAY_MS - Min inter-request delay (default: 500)
  RECALL_STEALTH_MAX_DELAY_MS - Max inter-request delay (default: 2000)
"""

from __future__ import annotations

import asyncio
import logging
import os
import random
from dataclasses import dataclass

import httpx

logger = logging.getLogger(__name__)

# ─── Browser Profiles ────────────────────────────────────────────────────────

CHROME_ACCEPT = (
    "text/html,application/xhtml+xml,application/xml;q=0.9,"
    "image/avif,image/webp,image/apng,*/*;q=0.8,"
    "application/signed-exchange;v=b3;q=0.7"
)
FIREFOX_ACCEPT = (
    "text/html,application/xhtml+xml,application/xml;q=0.9,"
    "image/avif,image/webp,*/*;q=0.8"
)
SAFARI_ACCEPT = "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"


@dataclass(frozen=True)
class BrowserProfile:
    user_agent: str
    sec_ch_ua: str | None
    sec_ch_ua_mobile: str
    sec_ch_ua_platform: str
    accept: str
    accept_language: str


BROWSER_PROFILES: list[BrowserProfile] = [
    BrowserProfile(
        user_agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
        sec_ch_ua='"Chromium";v="122", "Not(A:Brand";v="24", "Google Chrome";v="122"',
        sec_ch_ua_mobile="?0",
        sec_ch_ua_platform='"macOS"',
        accept=CHROME_ACCEPT,
        accept_language="en-US,en;q=0.9",
    ),
    BrowserProfile(
        user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36",
        sec_ch_ua='"Not A(Brand";v="99", "Google Chrome";v="121", "Chromium";v="121"',
        sec_ch_ua_mobile="?0",
        sec_ch_ua_platform='"Windows"',
        accept=CHROME_ACCEPT,
        accept_language="en-US,en;q=0.9",
    ),
    BrowserProfile(
        user_agent="Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
        sec_ch_ua='"Chromium";v="122", "Not(A:Brand";v="24", "Google Chrome";v="122"',
        sec_ch_ua_mobile="?0",
        sec_ch_ua_platform='"Linux"',
        accept=CHROME_ACCEPT,
        accept_language="en-US,en;q=0.9",
    ),
    BrowserProfile(
        user_agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:123.0) Gecko/20100101 Firefox/123.0",
        sec_ch_ua=None,
        sec_ch_ua_mobile="?0",
        sec_ch_ua_platform='"macOS"',
        accept=FIREFOX_ACCEPT,
        accept_language="en-US,en;q=0.5",
    ),
    BrowserProfile(
        user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:123.0) Gecko/20100101 Firefox/123.0",
        sec_ch_ua=None,
        sec_ch_ua_mobile="?0",
        sec_ch_ua_platform='"Windows"',
        accept=FIREFOX_ACCEPT,
        accept_language="en-US,en;q=0.5",
    ),
    BrowserProfile(
        user_agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.3 Safari/605.1.15",
        sec_ch_ua=None,
        sec_ch_ua_mobile="?0",
        sec_ch_ua_platform='"macOS"',
        accept=SAFARI_ACCEPT,
        accept_language="en-US,en;q=0.9",
    ),
    BrowserProfile(
        user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36 Edg/122.0.0.0",
        sec_ch_ua='"Chromium";v="122", "Not(A:Brand";v="24", "Microsoft Edge";v="122"',
        sec_ch_ua_mobile="?0",
        sec_ch_ua_platform='"Windows"',
        accept=CHROME_ACCEPT,
        accept_language="en-US,en;q=0.9",
    ),
]

_profile_counter = 0


def next_profile() -> BrowserProfile:
    """Round-robin profile selection."""
    global _profile_counter  # noqa: PLW0603
    idx = _profile_counter % len(BROWSER_PROFILES)
    _profile_counter += 1
    return BROWSER_PROFILES[idx]


def random_profile() -> BrowserProfile:
    """Random profile selection."""
    return random.choice(BROWSER_PROFILES)


# ─── Header Builders ─────────────────────────────────────────────────────────

def browser_headers(
    profile: BrowserProfile,
    referer: str | None = None,
) -> dict[str, str]:
    """Build a full set of browser-like headers from a profile."""
    headers: dict[str, str] = {
        "User-Agent": profile.user_agent,
        "Accept": profile.accept,
        "Accept-Language": profile.accept_language,
        "Accept-Encoding": "gzip, deflate, br",
        "DNT": "1",
        "Connection": "keep-alive",
        "Upgrade-Insecure-Requests": "1",
        "Sec-Fetch-Dest": "document",
        "Sec-Fetch-Mode": "navigate",
        "Sec-Fetch-User": "?1",
    }

    if profile.sec_ch_ua:
        headers["Sec-CH-UA"] = profile.sec_ch_ua
        headers["Sec-CH-UA-Mobile"] = profile.sec_ch_ua_mobile
        headers["Sec-CH-UA-Platform"] = profile.sec_ch_ua_platform

    if referer:
        headers["Referer"] = referer
        headers["Sec-Fetch-Site"] = "cross-site"
    else:
        headers["Sec-Fetch-Site"] = "none"

    return headers


def search_headers(
    profile: BrowserProfile,
    search_engine_referer: str | None = None,
) -> dict[str, str]:
    """Browser headers with search-engine-specific additions."""
    headers = browser_headers(profile, referer=None)
    headers["Cache-Control"] = "no-cache"
    headers["Pragma"] = "no-cache"

    if search_engine_referer:
        headers["Referer"] = search_engine_referer
        headers["Sec-Fetch-Site"] = "same-origin"

    return headers


# ─── Random Delay ─────────────────────────────────────────────────────────────

async def random_delay(default_min_ms: int = 500, default_max_ms: int = 2000) -> None:
    """Async sleep for a random duration to mimic human browsing."""
    min_ms = int(os.environ.get("RECALL_STEALTH_MIN_DELAY_MS", default_min_ms))
    max_ms = int(os.environ.get("RECALL_STEALTH_MAX_DELAY_MS", default_max_ms))
    max_ms = max(max_ms, min_ms)

    delay_s = random.uniform(min_ms / 1000, max_ms / 1000)
    logger.debug("Stealth: inter-request delay %.0fms", delay_s * 1000)
    await asyncio.sleep(delay_s)


# ─── Proxy Pool ───────────────────────────────────────────────────────────────

def _load_proxy_pool() -> list[str]:
    single = os.environ.get("RECALL_PROXY_URL", "").strip()
    if single:
        logger.info("Stealth: loaded single proxy from RECALL_PROXY_URL")
        return [single]

    multi = os.environ.get("RECALL_PROXY_URLS", "").strip()
    if multi:
        proxies = [p.strip() for p in multi.split(",") if p.strip()]
        if proxies:
            logger.info("Stealth: loaded %d proxies from RECALL_PROXY_URLS", len(proxies))
            return proxies

    return []


_PROXY_POOL: list[str] = _load_proxy_pool()


def random_proxy_url() -> str | None:
    """Pick a random proxy from the configured pool (or None)."""
    if not _PROXY_POOL:
        return None
    return random.choice(_PROXY_POOL)


def get_proxy_url() -> str | None:
    """Get a single proxy URL for httpx (compatible with all httpx versions)."""
    return random_proxy_url()


# ─── FlareSolverr Client ─────────────────────────────────────────────────────

class FlareSolverrClient:
    """Client for the FlareSolverr proxy that solves JS anti-bot challenges."""

    def __init__(self, endpoint: str):
        self.endpoint = endpoint.rstrip("/")
        self._client = httpx.AsyncClient(timeout=120.0)
        logger.info("Stealth: FlareSolverr client -> %s", self.endpoint)

    @classmethod
    def from_env(cls) -> FlareSolverrClient | None:
        url = os.environ.get("RECALL_FLARESOLVERR_URL", "").strip()
        if not url:
            return None
        return cls(url)

    async def solve(self, url: str, timeout_ms: int = 60000) -> str:
        """Send URL to FlareSolverr and return rendered HTML."""
        logger.debug("FlareSolverr: solving challenge for %s", url)
        resp = await self._client.post(
            f"{self.endpoint}/v1",
            json={"cmd": "request.get", "url": url, "maxTimeout": timeout_ms},
        )
        resp.raise_for_status()

        data = resp.json()
        if data.get("status") != "ok":
            msg = f"FlareSolverr status: {data.get('status')}"
            raise RuntimeError(msg)

        solution = data.get("solution")
        if not solution or "response" not in solution:
            msg = "FlareSolverr returned no solution"
            raise RuntimeError(msg)

        return solution["response"]


_FLARESOLVERR: FlareSolverrClient | None = FlareSolverrClient.from_env()


def flaresolverr() -> FlareSolverrClient | None:
    """Get the global FlareSolverr client (or None if unconfigured)."""
    return _FLARESOLVERR


# ─── Stealth HTTP Client Builder ─────────────────────────────────────────────

def build_stealth_httpx_client(
    timeout: float = 30.0,
    follow_redirects: bool = True,
) -> httpx.AsyncClient:
    """Create an httpx.AsyncClient with stealth features (cookies, proxy)."""
    kwargs: dict = {
        "timeout": timeout,
        "follow_redirects": follow_redirects,
        "limits": httpx.Limits(max_keepalive_connections=5, max_connections=10),
    }

    proxy_url = get_proxy_url()
    if proxy_url:
        kwargs["proxy"] = proxy_url

    return httpx.AsyncClient(**kwargs)

File: src/app/src/src/shared/utils/modules/stealth.rs
//! Stealth HTTP utilities for anti-bot evasion.
//!
//! Provides browser-like request profiles, user-agent rotation, proxy support,
//! random delays, and FlareSolverr integration to avoid detection.
//!
//! # Configuration (Environment Variables)
//!
//! - `RECALL_PROXY_URL`: Single residential proxy URL (e.g., `http://user:pass@proxy:8080`)
//! - `RECALL_PROXY_URLS`: Comma-separated proxy URLs for rotation
//! - `RECALL_FLARESOLVERR_URL`: FlareSolverr endpoint (e.g., `http://localhost:8191`)
//! - `RECALL_STEALTH_MIN_DELAY_MS`: Minimum inter-request delay (default: 500)
//! - `RECALL_STEALTH_MAX_DELAY_MS`: Maximum inter-request delay (default: 2000)

use once_cell::sync::Lazy;
use rand::Rng;
use reqwest::header::{
    HeaderMap, HeaderName, HeaderValue, ACCEPT, ACCEPT_LANGUAGE, CACHE_CONTROL, CONNECTION,
    REFERER, USER_AGENT,
};
use serde::Deserialize;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::time::Duration;
use tokio::time::sleep;
use tracing::{debug, info, warn};

/// Consistent browser profile pairing User-Agent with matching Sec-CH-UA headers.
/// Anti-bot systems flag mismatched UA/client-hint combinations.
pub struct BrowserProfile {
    pub user_agent: &'static str,
    pub sec_ch_ua: Option<&'static str>,
    pub sec_ch_ua_mobile: &'static str,
    pub sec_ch_ua_platform: &'static str,
    pub accept: &'static str,
    pub accept_language: &'static str,
}

const CHROME_ACCEPT: &str = "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7";
const FIREFOX_ACCEPT: &str =
    "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8";
const SAFARI_ACCEPT: &str = "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8";

const BROWSER_PROFILES: &[BrowserProfile] = &[
    BrowserProfile {
        user_agent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
        sec_ch_ua: Some("\"Chromium\";v=\"122\", \"Not(A:Brand\";v=\"24\", \"Google Chrome\";v=\"122\""),
        sec_ch_ua_mobile: "?0",
        sec_ch_ua_platform: "\"macOS\"",
        accept: CHROME_ACCEPT,
        accept_language: "en-US,en;q=0.9",
    },
    BrowserProfile {
        user_agent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36",
        sec_ch_ua: Some("\"Not A(Brand\";v=\"99\", \"Google Chrome\";v=\"121\", \"Chromium\";v=\"121\""),
        sec_ch_ua_mobile: "?0",
        sec_ch_ua_platform: "\"Windows\"",
        accept: CHROME_ACCEPT,
        accept_language: "en-US,en;q=0.9",
    },
    BrowserProfile {
        user_agent: "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
        sec_ch_ua: Some("\"Chromium\";v=\"122\", \"Not(A:Brand\";v=\"24\", \"Google Chrome\";v=\"122\""),
        sec_ch_ua_mobile: "?0",
        sec_ch_ua_platform: "\"Linux\"",
        accept: CHROME_ACCEPT,
        accept_language: "en-US,en;q=0.9",
    },
    BrowserProfile {
        user_agent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        sec_ch_ua: Some("\"Not_A Brand\";v=\"8\", \"Chromium\";v=\"120\", \"Google Chrome\";v=\"120\""),
        sec_ch_ua_mobile: "?0",
        sec_ch_ua_platform: "\"Windows\"",
        accept: CHROME_ACCEPT,
        accept_language: "en-US,en;q=0.9",
    },
    BrowserProfile {
        user_agent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:123.0) Gecko/20100101 Firefox/123.0",
        sec_ch_ua: None,
        sec_ch_ua_mobile: "?0",
        sec_ch_ua_platform: "\"macOS\"",
        accept: FIREFOX_ACCEPT,
        accept_language: "en-US,en;q=0.5",
    },
    BrowserProfile {
        user_agent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:123.0) Gecko/20100101 Firefox/123.0",
        sec_ch_ua: None,
        sec_ch_ua_mobile: "?0",
        sec_ch_ua_platform: "\"Windows\"",
        accept: FIREFOX_ACCEPT,
        accept_language: "en-US,en;q=0.5",
    },
    BrowserProfile {
        user_agent: "Mozilla/5.0 (X11; Linux x86_64; rv:123.0) Gecko/20100101 Firefox/123.0",
        sec_ch_ua: None,
        sec_ch_ua_mobile: "?0",
        sec_ch_ua_platform: "\"Linux\"",
        accept: FIREFOX_ACCEPT,
        accept_language: "en-US,en;q=0.5",
    },
    BrowserProfile {
        user_agent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.3 Safari/605.1.15",
        sec_ch_ua: None,
        sec_ch_ua_mobile: "?0",
        sec_ch_ua_platform: "\"macOS\"",
        accept: SAFARI_ACCEPT,
        accept_language: "en-US,en;q=0.9",
    },
    BrowserProfile {
        user_agent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36 Edg/122.0.0.0",
        sec_ch_ua: Some("\"Chromium\";v=\"122\", \"Not(A:Brand\";v=\"24\", \"Microsoft Edge\";v=\"122\""),
        sec_ch_ua_mobile: "?0",
        sec_ch_ua_platform: "\"Windows\"",
        accept: CHROME_ACCEPT,
        accept_language: "en-US,en;q=0.9",
    },
    BrowserProfile {
        user_agent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36 Edg/122.0.0.0",
        sec_ch_ua: Some("\"Chromium\";v=\"122\", \"Not(A:Brand\";v=\"24\", \"Microsoft Edge\";v=\"122\""),
        sec_ch_ua_mobile: "?0",
        sec_ch_ua_platform: "\"macOS\"",
        accept: CHROME_ACCEPT,
        accept_language: "en-US,en;q=0.9",
    },
];

// ─── Profile Rotation ───────────────────────────────────────────────────────

pub struct ProfileRotator {
    counter: AtomicUsize,
}

impl ProfileRotator {
    pub const fn new() -> Self {
        Self {
            counter: AtomicUsize::new(0),
        }
    }

    #[allow(clippy::indexing_slicing)]
    pub fn next_profile(&self) -> &'static BrowserProfile {
        let idx = self.counter.fetch_add(1, Ordering::Relaxed) % BROWSER_PROFILES.len();
        &BROWSER_PROFILES[idx]
    }

    #[allow(clippy::indexing_slicing)]
    pub fn random_profile(&self) -> &'static BrowserProfile {
        let mut rng = rand::thread_rng();
        let idx = rng.gen_range(0..BROWSER_PROFILES.len());
        &BROWSER_PROFILES[idx]
    }
}

static PROFILE_ROTATOR: Lazy<ProfileRotator> = Lazy::new(ProfileRotator::new);
static PROXY_POOL: Lazy<Option<Vec<String>>> = Lazy::new(load_proxy_pool);
static FLARESOLVERR: Lazy<Option<FlareSolverrClient>> = Lazy::new(FlareSolverrClient::from_env);

pub fn next_profile() -> &'static BrowserProfile {
    PROFILE_ROTATOR.next_profile()
}

pub fn random_profile() -> &'static BrowserProfile {
    PROFILE_ROTATOR.random_profile()
}

// ─── Header Builders ────────────────────────────────────────────────────────

/// Full set of browser-like headers for a navigation request.
///
/// When `referer` is supplied, `Sec-Fetch-Site` is set to `cross-site`
/// (mimicking a link click). Otherwise it is `none` (direct navigation).
pub fn browser_headers(profile: &BrowserProfile, referer: Option<&str>) -> HeaderMap {
    let mut headers = HeaderMap::with_capacity(14);

    headers.insert(USER_AGENT, HeaderValue::from_static(profile.user_agent));
    headers.insert(ACCEPT, HeaderValue::from_static(profile.accept));
    headers.insert(
        ACCEPT_LANGUAGE,
        HeaderValue::from_static(profile.accept_language),
    );
    headers.insert(
        HeaderName::from_static("dnt"),
        HeaderValue::from_static("1"),
    );
    headers.insert(CONNECTION, HeaderValue::from_static("keep-alive"));
    headers.insert(
        HeaderName::from_static("upgrade-insecure-requests"),
        HeaderValue::from_static("1"),
    );
    headers.insert(
        HeaderName::from_static("sec-fetch-dest"),
        HeaderValue::from_static("document"),
    );
    headers.insert(
        HeaderName::from_static("sec-fetch-mode"),
        HeaderValue::from_static("navigate"),
    );
    headers.insert(
        HeaderName::from_static("sec-fetch-user"),
        HeaderValue::from_static("?1"),
    );

    if let Some(ch_ua) = profile.sec_ch_ua {
        headers.insert(
            HeaderName::from_static("sec-ch-ua"),
            HeaderValue::from_static(ch_ua),
        );
        headers.insert(
            HeaderName::from_static("sec-ch-ua-mobile"),
            HeaderValue::from_static(profile.sec_ch_ua_mobile),
        );
        headers.insert(
            HeaderName::from_static("sec-ch-ua-platform"),
            HeaderValue::from_static(profile.sec_ch_ua_platform),
        );
    }

    if let Some(ref_url) = referer {
        if let Ok(val) = HeaderValue::from_str(ref_url) {
            headers.insert(REFERER, val);
            headers.insert(
                HeaderName::from_static("sec-fetch-site"),
                HeaderValue::from_static("cross-site"),
            );
        } else {
            headers.insert(
                HeaderName::from_static("sec-fetch-site"),
                HeaderValue::from_static("none"),
            );
        }
    } else {
        headers.insert(
            HeaderName::from_static("sec-fetch-site"),
            HeaderValue::from_static("none"),
        );
    }

    headers
}

/// Search-engine request headers: adds `Cache-Control: no-cache` and uses
/// `same-origin` for Sec-Fetch-Site when a search-engine referer is given.
pub fn search_headers(profile: &BrowserProfile, search_engine_referer: Option<&str>) -> HeaderMap {
    let mut headers = browser_headers(profile, None);

    headers.insert(CACHE_CONTROL, HeaderValue::from_static("no-cache"));
    headers.insert(
        HeaderName::from_static("pragma"),
        HeaderValue::from_static("no-cache"),
    );

    if let Some(ref_url) = search_engine_referer {
        if let Ok(val) = HeaderValue::from_str(ref_url) {
            headers.insert(REFERER, val);
            headers.insert(
                HeaderName::from_static("sec-fetch-site"),
                HeaderValue::from_static("same-origin"),
            );
        }
    }

    headers
}

// ─── Random Delay ───────────────────────────────────────────────────────────

/// Random delay between requests to mimic human browsing cadence.
/// Bounds can be overridden via `RECALL_STEALTH_MIN_DELAY_MS` / `RECALL_STEALTH_MAX_DELAY_MS`.
pub async fn random_delay(default_min_ms: u64, default_max_ms: u64) {
    let delay_ms = {
        let min_ms = std::env::var("RECALL_STEALTH_MIN_DELAY_MS")
            .ok()
            .and_then(|v| v.parse().ok())
            .unwrap_or(default_min_ms);
        let max_ms = std::env::var("RECALL_STEALTH_MAX_DELAY_MS")
            .ok()
            .and_then(|v| v.parse().ok())
            .unwrap_or(default_max_ms)
            .max(min_ms);
        rand::thread_rng().gen_range(min_ms..=max_ms)
    };
    debug!(delay_ms, "Stealth: inter-request delay");
    sleep(Duration::from_millis(delay_ms)).await;
}

// ─── Proxy Pool ─────────────────────────────────────────────────────────────

fn load_proxy_pool() -> Option<Vec<String>> {
    if let Ok(url) = std::env::var("RECALL_PROXY_URL") {
        let url = url.trim().to_string();
        if !url.is_empty() {
            info!("Stealth: loaded single proxy from RECALL_PROXY_URL");
            return Some(vec![url]);
        }
    }
    if let Ok(urls) = std::env::var("RECALL_PROXY_URLS") {
        let proxies: Vec<String> = urls
            .split(',')
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .collect();
        if !proxies.is_empty() {
            info!(
                count = proxies.len(),
                "Stealth: loaded proxy pool from RECALL_PROXY_URLS"
            );
            return Some(proxies);
        }
    }
    debug!("Stealth: no proxy configured");
    None
}

/// Pick a random proxy URL from the configured pool (if any).
#[allow(clippy::indexing_slicing)]
pub fn random_proxy_url() -> Option<&'static str> {
    PROXY_POOL.as_ref().map(|proxies| {
        let mut rng = rand::thread_rng();
        let idx = rng.gen_range(0..proxies.len());
        proxies[idx].as_str()
    })
}

/// Get the FlareSolverr client (if `RECALL_FLARESOLVERR_URL` is set).
pub fn flaresolverr() -> Option<&'static FlareSolverrClient> {
    FLARESOLVERR.as_ref()
}

// ─── FlareSolverr Client ────────────────────────────────────────────────────

/// Proxy server that solves Cloudflare / JS anti-bot challenges using a real
/// headless browser. See <https://github.com/FlareSolverr/FlareSolverr>.
pub struct FlareSolverrClient {
    endpoint: String,
    client: reqwest::Client,
    timeout_ms: u64,
}

#[derive(Deserialize)]
struct FlareSolverrResponse {
    status: String,
    solution: Option<FlareSolverrSolution>,
}

#[derive(Deserialize)]
struct FlareSolverrSolution {
    response: String,
    #[allow(dead_code)]
    status: u16,
    #[allow(dead_code)]
    url: String,
}

impl FlareSolverrClient {
    fn from_env() -> Option<Self> {
        let endpoint = std::env::var("RECALL_FLARESOLVERR_URL").ok()?;
        let endpoint = endpoint.trim().to_string();
        if endpoint.is_empty() {
            return None;
        }
        let client = reqwest::Client::builder()
            .timeout(Duration::from_secs(120))
            .build()
            .ok()?;

        info!(endpoint = %endpoint, "Stealth: FlareSolverr client initialized");
        Some(Self {
            endpoint,
            client,
            timeout_ms: 60000,
        })
    }

    /// Solve a JS challenge by delegating to FlareSolverr. Returns rendered HTML.
    pub async fn solve(&self, url: &str) -> std::result::Result<String, String> {
        debug!(url, "FlareSolverr: solving challenge");

        let body = serde_json::json!({
            "cmd": "request.get",
            "url": url,
            "maxTimeout": self.timeout_ms
        });

        let response = self
            .client
            .post(&format!("{}/v1", self.endpoint.trim_end_matches('/')))
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("FlareSolverr request failed: {}", e))?;

        if !response.status().is_success() {
            return Err(format!("FlareSolverr HTTP {}", response.status()));
        }

        let result: FlareSolverrResponse = response
            .json()
            .await
            .map_err(|e| format!("FlareSolverr parse failed: {}", e))?;

        if result.status != "ok" {
            return Err(format!("FlareSolverr status: {}", result.status));
        }

        result
            .solution
            .map(|s| s.response)
            .ok_or_else(|| "FlareSolverr returned no solution".to_string())
    }
}

// ─── Stealth Client Builder ─────────────────────────────────────────────────

/// Reqwest `ClientBuilder` with stealth features:
/// - Cookie jar for session persistence
/// - Proxy rotation (when configured)
///
/// Does **not** set `User-Agent` — use [`browser_headers`] per-request for rotation.
pub fn stealth_client_builder() -> reqwest::ClientBuilder {
    let mut builder = super::http_client::reqwest_client_builder().cookie_store(true);

    if let Some(proxy_url) = random_proxy_url() {
        match reqwest::Proxy::all(proxy_url) {
            Ok(proxy) => {
                debug!(proxy = %proxy_url, "Stealth: using proxy");
                builder = builder.proxy(proxy);
            }
            Err(e) => {
                warn!(proxy = %proxy_url, error = %e, "Stealth: invalid proxy URL, skipping");
            }
        }
    }

    builder
}

/// Infer the search-engine origin from a request URL for the Referer header.
pub fn search_referer_for_url(url: &str) -> Option<&'static str> {
    if url.contains("duckduckgo.com") {
        Some("https://duckduckgo.com/")
    } else if url.contains("bing.com") {
        Some("https://www.bing.com/")
    } else if url.contains("wikipedia.org") {
        Some("https://en.wikipedia.org/")
    } else {
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_profile_rotation_wraps_around() {
        let rotator = ProfileRotator::new();
        let first = rotator.next_profile().user_agent;
        for _ in 1..BROWSER_PROFILES.len() {
            let _ = rotator.next_profile();
        }
        let wrapped = rotator.next_profile().user_agent;
        assert_eq!(first, wrapped);
    }

    #[test]
    fn test_browser_headers_no_referer() {
        let profile = &BROWSER_PROFILES[0];
        let headers = browser_headers(profile, None);
        assert!(headers.get(USER_AGENT).is_some());
        assert_eq!(
            headers
                .get(HeaderName::from_static("sec-fetch-site"))
                .and_then(|v| v.to_str().ok()),
            Some("none")
        );
        assert!(headers.get(REFERER).is_none());
    }

    #[test]
    fn test_browser_headers_with_referer() {
        let profile = &BROWSER_PROFILES[0];
        let headers = browser_headers(profile, Some("https://www.google.com/"));
        assert_eq!(
            headers.get(REFERER).and_then(|v| v.to_str().ok()),
            Some("https://www.google.com/")
        );
        assert_eq!(
            headers
                .get(HeaderName::from_static("sec-fetch-site"))
                .and_then(|v| v.to_str().ok()),
            Some("cross-site")
        );
    }

    #[test]
    fn test_search_headers_same_origin() {
        let profile = &BROWSER_PROFILES[0];
        let headers = search_headers(profile, Some("https://duckduckgo.com/"));
        assert_eq!(
            headers
                .get(HeaderName::from_static("sec-fetch-site"))
                .and_then(|v| v.to_str().ok()),
            Some("same-origin")
        );
        assert!(headers.get(CACHE_CONTROL).is_some());
    }

    #[test]
    fn test_search_referer_detection() {
        assert_eq!(
            search_referer_for_url("https://html.duckduckgo.com/html/?q=test"),
            Some("https://duckduckgo.com/")
        );
        assert_eq!(
            search_referer_for_url("https://www.bing.com/search?q=test"),
            Some("https://www.bing.com/")
        );
        assert_eq!(search_referer_for_url("https://example.com"), None);
    }

    #[test]
    fn test_profiles_have_consistent_headers() {
        for profile in BROWSER_PROFILES {
            let ua_lower = profile.user_agent.to_ascii_lowercase();
            if ua_lower.contains("chrome") && !ua_lower.contains("firefox") {
                assert!(
                    profile.sec_ch_ua.is_some(),
                    "Chrome-based UA should have Sec-CH-UA: {}",
                    profile.user_agent
                );
            }
            if ua_lower.contains("firefox") {
                assert!(
                    profile.sec_ch_ua.is_none(),
                    "Firefox UA should not have Sec-CH-UA: {}",
                    profile.user_agent
                );
            }
        }
    }
}
